### ISSUE TITLE
segfaults on 64-bit windows

### ISSUE THREAD

**Issue by digama0** [metamath/metamath-exe#164]
I don't have a windows machine, but the cross compile for 64-bit windows on my machine looks like this:
```
$ cd build
$ make clean
$ make CC=x86_64-w64-mingw32-gcc
$ wine src/metamath.exe
0024:err:file:init_redirects /home/mario/.wine/dosdevices/c:/windows: No such file or directory
002c:err:wineboot:main Cannot set the dir to L"C:\\windows" (2)
002c:err:wineboot:process_run_key Error running cmd L"C:\\windows\\system32\\winemenubuilder.exe -a -r" (2).
wine: could not open working directory L"unix\\home\\mario\\Documents\\metamath\\metamath-exe\\build\\", starting in the Windows directory.
002c:err:shell:SHGetFolderPathAndSubDirW Failed to create directory L"C:\\users\\mario\\Application Data".
002c:err:shell:SHGetFolderPathAndSubDirW Failed to create directory L"C:\\users\\mario".
002c:err:shell:SHGetFolderPathAndSubDirW Failed to create directory L"C:\\users\\mario\\Local Settings\\Application Data".
wine: could not open working directory L"unix\\home\\mario\\Documents\\metamath\\metamath-exe\\build\\", starting in the Windows directory.
*** BUG #1501
Metamath - Version 0.199.pre 29-Jan-2022      wine: Unhandled page fault on read access to 0000000000000000 at address 00000001C8DFD2B0 (thread 0040), starting debugger...
0040:err:seh:start_debugger Couldn't start debugger L"winedbg --auto 60 60" (2)
Read the Wine Developers Guide on how to set up winedbg or another debugger
```
Most of the wine errors are harmless, but metamath immediately gives `BUG #1501` and then dereferences null shortly afterward. The 32-bit windows build works properly.

I believe the reason is because metamath-exe uses `long` as an equivalent to `size_t`, `ptrdiff_t` and the like, but in 64-bit windows `long` is 4 bytes while pointers and allocation sizes are 8 bytes, so there is a lot of invalid casting going around.

**Comment by creator1creeper1** [comment]
Thank you for nothing this issue. Compiling with `clang m*.c -o ../bin/metamath.exe -O3 -funroll-loops -finline-functions -fomit-frame-pointer -m32` (note the `-m32`) works on my machine.

**Comment by vince-gonzalez** [comment]
Confirmed on native Windows 11 x64 rather than wine — 0.199.pre built from `master` with `zig cc -target x86_64-windows-gnu`. `*** BUG #1501` at startup, then a segfault on the first `read`.

The cause is the layout of `g_PntrNull` against `pntrLen()`.

`pntrLen()` reads the `long` immediately before the payload:

```c
long pntrLen(const pntrString *s) {
  return ((((const long *)s)[-1] - (long)(sizeof(pntrString)))
      / (long)(sizeof(pntrString)));
}
```

`poolMalloc()` builds that by hand — `ptr = (long *)ptr + 3;` then `((long *)ptr)[-1] = size;` — so for a pool allocation the read lands on `actualSize`. `NULL_PNTRSTRING` instead points into a struct, where the compiler chooses the layout:

```c
struct nullPntrStruct {
  long poolLoc;
  long allocSize;
  long actualSize;
  pntrString nullElement;
};
```

Where a pointer is more strictly aligned than a `long`, padding is inserted between `actualSize` and `nullElement`, and `pntrLen()` reads the padding. LP64 has both at 8 and ILP32 has both at 4, so neither pads — which is why `-m32` works. **LLP64, i.e. 64-bit Windows, has `long` at 4 and pointers at 8, so four bytes of padding sit exactly where the length belongs.** `print2()` then takes the `pntrLen(backBuffer)` branch on first entry and reports 1501.

Reproduction without metamath — same struct, same `pntrLen`, compiled to both targets:

```
                           x86_64-windows-gnu   x86-windows-gnu
sizeof(long)                                4                 4
sizeof(pntrString)                          8                 4
offsetof(actualSize)                        8                 8
offsetof(nullElement)                      16                12
padding before nullElement                  4                 0
value pntrLen reads at [-1]                 0                 4
pntrLen(NULL_PNTRSTRING)                   -1                 0
```

`mmdata.h` already carries a `\bug` note on that struct — "The C standard does not require a long having the same size as a void\*" — so this is that documented concern being reachable, not a new one.

I haven't proposed a patch. Padding the struct would fix the null block while leaving `poolMalloc()` handing back payloads aligned only to `long`, and choosing between that and aligning the pool payload properly looks like a decision for the maintainers.

**Comment by tirix** [comment]
Do we have access to `__attribute__((__packed__))` in the C version we are using ?

**Comment by benjub** [comment]
> Do we have access to `__attribute__((__packed__))` in the C version we are using ?

That looks a bit hacky. There should be an easier way to access the component we want, using `offsetof(struct nullPntrStruct, nullElement)`.

**Comment by vince-gonzalez** [comment]
`packed` is available in gcc and clang (it's a GNU extension — MSVC would want `#pragma pack`, though the autotools build doesn't go there), and it does fix this. Marking only `nullPntrStruct` and building for `x86_64-windows-gnu`:

```
                             before    packed
offsetof(actualSize)              8         8
offsetof(nullElement)            16        12
padding before nullElement        4         0
pntrLen(NULL_PNTRSTRING)         -1         0
```

That binary starts clean and reads set.mm. Running `write source set.mm /rewrap` under it and under a 32-bit build of the same commit gives byte-identical output — 52,061,323 bytes, 250,863 statements from each.

Two things I'd rather flag than argue.

It warns `-Waddress-of-packed-member`, since `nullElement` becomes 4-aligned instead of 8. That is the alignment `poolMalloc()` already returns, though — `ptr = (long *)ptr + 3` puts the payload 12 bytes in on LLP64, so pool-allocated `pntrString`s are 4-aligned there too. Packing the struct makes the static match the pool rather than introducing a new condition. Whether that's a reason to accept it or a reason to align both is a call I'd leave to you.

On `offsetof`, I may be misreading the suggestion. `pntrLen()` takes a bare `const pntrString *` and is called on both pool blocks and `NULL_PNTRSTRING`, so from inside it there's no containing struct to take an offset in. A guard like `if (s == NULL_PNTRSTRING) return 0;` would sidestep the layout and is fully portable — but `pntrAllocLen()` reads `[-2]` and `pntrZapLen()` reads `[-3]` and `[-1]` on the same assumption, so it would want the same treatment in three places. If the idea was to use `offsetof` at the definition site to position the fields, I'd like to see it; I couldn't find a portable way without conditional compilation.

Happy to open a PR for whichever direction you prefer, or to leave it with you.

**Comment by tirix** [comment]
> It warns `-Waddress-of-packed-member`, since `nullElement` becomes 4-aligned instead of 8. That is the alignment `poolMalloc()` already returns, though — `ptr = (long *)ptr + 3` puts the payload 12 bytes in on LLP64, so pool-allocated `pntrString`s are 4-aligned there too. Packing the struct makes the static match the pool rather than introducing a new condition. Whether that's a reason to accept it or a reason to align both is a call I'd leave to you.

The quick and easy solution (1) for this could be to add a dummy element to the struct, to make `nullElement` 8-bytes aligned, as in:
```
struct nullPntrStruct {
  long dummy; /* ensure that nullElement is 8-bytes aligned in 64-bit Windows */
  long poolLoc;
  long allocSize;
  long actualSize;
  pntrString nullElement;
};
```
I agree this looks (even more?) hacky. 

An alternative (2) might be to use `pntrString` to access the `actualSize` instead of `long`:
```
long pntrLen(const pntrString *s) {
  return ((s[-1] - (long)(sizeof(pntrString)))
      / (long)(sizeof(pntrString)));
}
```
Since `sizeof(long)` and `sizeof(pntrString)` are the same on other platforms, this should not impact other platforms. However for Windows 64 bits we would point to the right element. The same change has to be done _at least_ in `pntrAllocLen`, `nmbrLen`, `nmbrAllocLen`.

Another alternative (3) could be to call `poolMalloc` to initialize the `g_PntrNull` variable. That would have to be done right after the pool is initialized, before any use of that variable. I'm not fully sure of the implications. This way, `g_PntrNull` would follow exactly the same rules as any other string allocated with `poolMalloc`.

Yet another alternative (4), would be to follow [this remark](https://github.com/metamath/metamath-exe/pull/9#discussion_r776613820) from @digama0 : rewrite `nmbrLen` , `pntrLen`, `nmbrAllocLen‎`, etc. and _all places where the assumption is made that `poolMalloc` was used_ using a calculation involving the sizes of the pointers. That might be cumbersome, replacing `-1` with something like `-(sizeof(pntrString) / sizeof(long))` or defining a macro and using it everywhere.

Alternative (1) is the easiest since it only impact one place. I think I prefer (2), but we'd have to find all places to update.

In the end, all are somewhat hacky, but I think that's part of the essence of coding in C and dealing with low level memory allocation and alignment.

**Comment by benjub** [comment]
(1) looks best to me for an easy fix. It now relies on
```
_Static_assert(offsetof(struct nullPntrStruct, nullElement)
    == offsetof(struct nullPntrStruct, actualSize) + sizeof(long),
                "padding before nullElement; [-1] would not read actualSize");
```
(and same with `nullNmbrStruct`).

Also:
```
long pntrLen(const pntrString *s) {
  if (s == NULL_PNTRSTRING) return 0;
  ...
```
"hides" that specific bug. That early exit could still be added independently of the above.

**Comment by tirix** [comment]
Maybe @wlammen has an opinion here?

If no other direction is proposed, @vince-gonzalez, could you propose a PR?

**Comment by vince-gonzalez** [comment]
I tested both against a poolMalloc-shaped block as well as the static, since a fix has to hold for both.

```
                                       len_long   len_pntr
static, current struct   (want 0)            -1          0
static, option 1 struct  (want 0)             0          0
pool, allocSize == actualSize (want 4)        4          4
pool, allocSize != actualSize (want 4)        4         39
```

Option 1 is correct in all four.

Option 2 has a problem on the last row. `s[-1]` on a `pntrString *` reads 8 bytes on LLP64, spanning `allocSize` and `actualSize`, and the cast back to `long` truncates to `allocSize`. Those are equal for a fresh allocation, so it looks right. `poolFixedMalloc()` keeps the larger `allocSize` when it reuses a block, and then they differ: above, `actualSize` says 4 elements and it returns 39. Both pass on 32-bit, so this only appears on Win64 with a reused block.

The `_Static_assert` on `offsetof` reads well to me. It would have caught the original at compile time on Windows rather than at the first `print2()`.

**Comment by wlammen** [comment]
If possible, reorder the members of the  struct according to size, so nullElement becomes the first one.

Not possible, must be last, because it represents data following a header.

**Comment by wlammen** [comment]
Use long long for the integer members of nullPntrStruct, when you are compiling for a 64 bit architecture. The type long long is 64 bit on all major 64 bit systems, and differs from long only on Windows. See also the \bug entry in the comment to nullPntrStruct, that exactly addresses the problem here. But do not use long long on 32 bit systems.

[metamath executable_ Block of memory.pdf](https://github.com/user-attachments/files/31386602/metamath.executable_.Block.of.memory.pdf) shows that aligning, and only of one member is not sufficient. 

Clang and gcc support `__SIZEOF_POINTER__` for years now.  MSVC should check for `_WIN64` instead.
Add at the beginning of mmdata.h
```
#if defined(__SIZEOF_POINTER__) && __SIZEOF_POINTER__ >= 8
  // 64-bit executable
  typedef long long pntrInt;
#else
  // 32-bit executable
  typedef long pntrInt;
#endif
```
and then use this type to define the struct:
```
struct nullPntrStruct {
  pntrInt poolLoc;
  pntrInt allocSize;
  pntrInt actualSize;
  pntrString nullElement;
};
```

Finally you have to check the assignments to the members, and use 1LL instead of 1 or 1L, for example. Let the compiler lead you to accesses by temporarily suffixing the member names with an underscore,  and repeatedly start compilation, until all accesses are pointed out and checked.

Revert the name changes afterwards, and don't forget to update the comment to nullPntrString.

**Comment by vince-gonzalez** [comment]
I tested the `pntrInt` proposal before proposing a PR, and it needs a wider change than the struct. Table first, then why.

Target x86_64-windows-gnu, `void*` 8, `long` 4, `long long` 8, `intptr_t` 8. The check is `pntrLen(&g_PntrNull.nullElement)`, which must be 0. "shipped" is `pntrLen` exactly as it is in mmdata.c:2675, casting to `const long *`. "widened" is the same function with the cast changed to the widened header type.

```
                    hdr  off              shipped      widened
  current            12   16  MISALIGN     -1 WRONG      0 ok
  packed             12   12  aligned       0 ok         2147483647 WRONG
  longlong           24   24  aligned      -1 WRONG      0 ok
  intptr             24   24  aligned      -1 WRONG      0 ok
```

On x86-windows-gnu all four read 0 under both accessors, so nothing here regresses 32-bit.

The reason `longlong` and `intptr` read -1 under the shipped accessor is that `pntrLen` and `pntrAllocLen` cast to `long` unconditionally, so a widened header is read half at a time. That much is fixable by widening those two functions. The larger problem is that they are not the only readers.

`poolFixedMalloc` defines the header layout (mmdata.c:400-407):

```c
ptr = malloc( 3 * sizeof(long) + (size_t)size);
ptr = (long *)ptr + 3;
((long *)ptr)[-1] = size;   // Actual size
((long *)ptr)[-2] = size;   // Allocated size
((long *)ptr)[-3] = -1;     // Location in memUsedPool
```

The header of every pooled block is exactly `3 * sizeof(long)`, written as three longs. `nullPntrStruct` and `nullNmbrStruct` exist to mimic that layout for a statically allocated block, which is why they must agree with it. Widening the two structs and not the allocator makes the static instances 24-byte-header objects among 12-byte-header ones, and the accessors cannot be right for both at once.

Making it consistent means widening `poolFixedMalloc`, `poolMalloc`, `poolFree`, `getPoolStats` and the rest — I count about 45 `((long *)ptr)[-1]` / `[-2]` / `[-3]` sites in mmdata.c. Those serve `nmbrString` as well as `pntrString`, since both use the same pool, so the change reaches `nullNmbrStruct`, `nmbrLen`, `nmbrAllocLen` and the nmbr copy paths too. That is a real refactor rather than a fix for this issue, though it is the one the `\bug` note in the comment is pointing at.

Worth noting that `nullNmbrStruct` is not affected on LLP64. Its last member is `nmbrString`, which is `long`, so the struct is four 4-byte members with no padding and the element lands at offset 12, matching the pool header. Only `nullPntrStruct` breaks, because `pntrString` is `void*` and its 8-byte alignment inserts 4 bytes of padding after `actualSize`. That is why the bug is specific to pntrStrings on Windows.

So the two directions are not the same size of change, and they are not composable — `packed` is correct only while the accessors stay `long`, and widening is correct only once every reader is widened.

If the preference is the small fix now, I will open a PR with the packed struct plus the `_Static_assert` on `offsetof` that @benjub suggested, which turns this from a runtime segfault into a compile error on any platform where the layouts diverge. If the preference is `pntrInt` done properly, I am happy to do that instead, but it should be its own issue since it changes the pool allocator and touches nmbrString, and it wants review from whoever knows the pool code best.

Test program and the three cross-compiled binaries are reproducible with `zig cc -target x86_64-windows-gnu`; I can attach it if useful.

**Comment by wlammen** [comment]
Thanks for doing this. I have no access to Windows computers, so I could not do it myself.

**Comment by david-a-wheeler** [comment]
Thanks, I don't have a Windows system handy for this. Be careful on the reorder; modern computers' compute speed is often domination by memory cache line interference. In an earlier "benign" change to a structure I ended up making Metamath-exe take 50% longer because I made the cacheline problems much worse.

I suggest timing the original & changed results.

**Comment by vince-gonzalez** [comment]
Timed it. The layout comparison answers it more directly, so that first:

```
  ILP32   x86-windows      sizeof 16, element at 12    unchanged
  LP64    x86_64-linux     sizeof 32, element at 24    unchanged
  LLP64   x86_64-windows   24 -> 20,  element 16 -> 12
```

The layout only moves on LLP64, which segfaults during start-up today, so there is no working "before" for it to be slower than. On everything that runs now the struct is byte for byte what it was.

Timing anyway. 32-bit, built with the `-O3 -funroll-loops -finline-functions -fomit-frame-pointer` from configure.ac, `verify proof *` over set.mm, ten runs of each alternating:

```
  master   min 4.63  median 5.00  mean 5.03
  patched  min 4.67  median 4.95  mean 4.95
```

Min to min +0.9%, median to median -1.2%. The sign changes with the statistic and both are inside the ~0.7 s spread within each build.

Worth adding that there is one instance of this struct, `g_PntrNull` at mmdata.c:103, rather than an array, so there is no cache-line density to lose. The struct with that property is `statement_struct`, which is what the existing `_Static_assert` at mmdata.c:34 pins at 256 bytes; this does not touch it.
